File: script/make/protocol/js/TaskProtocol.js
export default class TaskProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 11201: {

                return new DataView(view.buffer.slice(0, offset));
            }
            case 11202: {
                // extend
                while (view.byteLength < offset + 4) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // 任务ID
                view.setUint32(offset, data, false);
                offset = offset + 4;
                return new DataView(view.buffer.slice(0, offset));
            }
            case 11203: {
                // extend
                while (view.byteLength < offset + 4) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // 任务ID
                view.setUint32(offset, data, false);
                offset = offset + 4;
                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 11201: {
                // 任务列表
                const data = [];
                let dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--dataLength >= 0) {
                    // 
                    // 任务ID
                    const dataDataTaskId = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 当前数量
                    const dataDataNumber = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 是否领取奖励
                    const dataDataIsAward = view.getUint8(offset, false);
                    offset = offset + 1;
                    // object
                    const dataData = {"taskId": dataDataTaskId, "number": dataDataNumber, "isAward": dataDataIsAward};
                    // add
                    data.push(dataData);
                }
                return data;
            }
            case 11202: {
                // 
                // 结果
                const dataResultLength = view.getUint16(offset, false);
                offset = offset + 2;
                const dataResultArray = new Uint8Array(view.buffer.slice(offset, offset + dataResultLength));
                const dataResult = textDecoder.decode(dataResultArray);
                offset = offset + dataResultLength;
                // 
                // 任务ID
                const dataTaskTaskId = view.getUint32(offset, false);
                offset = offset + 4;
                // 当前数量
                const dataTaskNumber = view.getUint16(offset, false);
                offset = offset + 2;
                // 是否领取奖励
                const dataTaskIsAward = view.getUint8(offset, false);
                offset = offset + 1;
                // object
                const dataTask = {"taskId": dataTaskTaskId, "number": dataTaskNumber, "isAward": dataTaskIsAward};
                // object
                const data = {"result": dataResult, "task": dataTask};
                return data;
            }
            case 11203: {
                // 结果
                const dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                const dataArray = new Uint8Array(view.buffer.slice(offset, offset + dataLength));
                const data = textDecoder.decode(dataArray);
                offset = offset + dataLength;
                return data;
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}

File: script/make/protocol/lua/TaskProtocol.lua
TaskProtocol = {}

function TaskProtocol.encode(offset, protocol, data)
    if protocol == 11201 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 11202 then
        local offset = offset
        local table = {}
        -- 任务ID
        table[offset] = string.pack(">I4", data["taskId"])
        offset = offset + 1
        return table
    elseif protocol == 11203 then
        local offset = offset
        local table = {}
        -- 任务ID
        table[offset] = string.pack(">I4", data["taskId"])
        offset = offset + 1
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function TaskProtocol.decode(offset, protocol, data)
    if protocol == 11201 then
        local offset = offset
        -- 任务列表
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- Task
            -- 任务ID
            local taskId = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 当前数量
            local number = string.unpack(">I2", data, offset)
            offset = offset + 2
            -- 是否领取奖励
            local isAward = string.unpack(">I1", data, offset)
            offset = offset + 1
            -- object
            local task = {taskId = taskId, number = number, isAward = isAward}
            list[listIndex] = task
        end
        return {list = list}
    elseif protocol == 11202 then
        local offset = offset
        -- 结果
        local result = string.unpack(">s2", data, offset)
        offset = offset + 2 + string.len(result)
        -- Task
        -- 任务ID
        local taskId = string.unpack(">I4", data, offset)
        offset = offset + 4
        -- 当前数量
        local number = string.unpack(">I2", data, offset)
        offset = offset + 2
        -- 是否领取奖励
        local isAward = string.unpack(">I1", data, offset)
        offset = offset + 1
        -- object
        local task = {taskId = taskId, number = number, isAward = isAward}
        return {result = result, task = task}
    elseif protocol == 11203 then
        local offset = offset
        -- 结果
        local result = string.unpack(">s2", data, offset)
        offset = offset + 2 + string.len(result)
        return {result = result}
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end